Using `rotate: x` and `transform: rotate(x)` yields different behavior with SVG

graouts opened the pull request:
#### 540afd68832196a977584e05afc0a139f564063f
<pre>
Using `rotate: x` and `transform: rotate(x)` yields different behavior with SVG
<a href="https://bugs.webkit.org/show_bug.cgi?id=250387">https://bugs.webkit.org/show_bug.cgi?id=250387</a>

Reviewed by Simon Fraser.

Check whether any transform-related property changed in LegacyRenderSVGModelObject::styleDidChange()
to determine whether an SVG element&apos;s needs to update its transform. To support this we add a new
RenderStyle::affectsTransform() method which returns true when any of the RenderStyle data
structures related to the &quot;transform&quot;, &quot;scale&quot;, &quot;rotate&quot;, &quot;translate&quot; and &quot;offset&quot; CSS properties
are non-empty or defined.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-animation-on-svg-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-animation-on-svg-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-animation-on-svg.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/scale-animation-on-svg-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/scale-animation-on-svg-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/scale-animation-on-svg.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/translate-animation-on-svg-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/translate-animation-on-svg-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/translate-animation-on-svg.html: Added.
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::affectsTransform const):
* Source/WebCore/rendering/svg/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::styleDidChange):

Canonical link: <a href="https://commits.webkit.org/258882@main">https://commits.webkit.org/258882@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f88b229dcd167a9980ee4db4fc9603d3429d3c9d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103221 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12346 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/36216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112468 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172665 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107174 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13371 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3250 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95447 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/110668 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108995 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10278 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/36216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37906 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92089 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/36216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79635 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5749 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/36216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5920 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2864 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11909 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/36216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7668 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -712,6 +712,8 @@ class RenderStyle {
     ScaleTransformOperation* scale() const { return m_rareNonInheritedData->scale.get(); }
     TranslateTransformOperation* translate() const { return m_rareNonInheritedData->translate.get(); }
 
+    bool hasTransformOrRelated() const { return hasTransform() || offsetPath() || rotate() || scale() || translate(); }
```
"OrRelated" is a bit vague, and this is rather similar to `hasTransformRelatedProperty`. Also, `hasTransform()` checks `offsetPath()`. So maybe call this `hasTransformAffectingProperty` and have it call `hasTransform`.

graouts replied:
I wanted to stay clear of the term `property` because RenderStyle is not really about properties. It's not that big of a deal, but maybe `affectsTransform()` would be better?

As for `hasTransform()`, it's probably a mistake that it checks `offsetPath()`, or at least it conflates two things which shouldn't always be conflated. To wit, `computedTransform()` calls it but is only interested in knowing whether a transform is specified.

I'm going to start a task of auditing call sites of `RenderStyle::hasTransform()` to check whether it's right to call that method, or whether it should use the new `RenderStyle::affectsTransform()` method. I think there are oversights or errors related to the use of `RenderStyle::hasTransform()` as it stands.

smfr commented on Source/WebCore/rendering/svg/LegacyRenderSVGModelObject.cpp:
```diff
@@ -105,7 +105,7 @@ void LegacyRenderSVGModelObject::styleDidChange(StyleDifference diff, const Rend
     if (diff == StyleDifference::Layout) {
         setNeedsBoundariesUpdate();
-        if (style().hasTransform() || (oldStyle && oldStyle->hasTransform()))
+        if (style().hasTransformOrRelated() || (oldStyle && oldStyle->hasTransformOrRelated()))
```
Should this also check if the transform changed?

graouts replied:
I'll look into this, although I wanted to make a minimal change.

grorg commented on LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-animation-on-svg.html:
```diff
@@ -0,0 +1,42 @@
+
+@keyframes rotate-animation {
+    from { rotate: 0; }
+    to   { rotate: 180deg; }
```
How can you tell this animation happened? 180deg on a 100x100 rectangle is the same as not doing anything. Shouldn't you use 45deg?

grorg replied:
"looks the same"

graouts replied:
The `transform-origin` makes it so that it's distinguishable.

graouts committed:
Using `rotate: x` and `transform: rotate(x)` yields different behavior with SVG
https://bugs.webkit.org/show_bug.cgi?id=250387

Reviewed by Simon Fraser.

Check whether any transform-related property changed in LegacyRenderSVGModelObject::styleDidChange()
to determine whether an SVG element's needs to update its transform. To support this we add a new
RenderStyle::affectsTransform() method which returns true when any of the RenderStyle data
structures related to the "transform", "scale", "rotate", "translate" and "offset" CSS properties
are non-empty or defined.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-animation-on-svg-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-animation-on-svg-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-animation-on-svg.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/scale-animation-on-svg-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/scale-animation-on-svg-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/scale-animation-on-svg.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/translate-animation-on-svg-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/translate-animation-on-svg-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/translate-animation-on-svg.html: Added.
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::affectsTransform const):
* Source/WebCore/rendering/svg/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::styleDidChange):

Canonical link: https://commits.webkit.org/258882@main
